### ISSUE TITLE
Error when compiling worker code with gfortran 10

### ISSUE THREAD

**Issue by fjansson** [amusecode/amuse#648]
I recently updated my Arch Linux system and got gfortran 10.
Now I cannot install OMUSE workers anymore - specifically DALES.
I have amuse-framework-13.2.0 installed with pip and the current OMUSE installed from github.

```
mpif90 -g -O2 -fPIC -fdefault-real-8 -cpp -I/home/fjansson/work_F/cwi-code/omuse/src/omuse/community/dales/dales-build
/src/library_modules/ -I/home/fjansson/work_F/cwi-code/omuse_env/share/amuse/lib/stopcond -I/home/fjansson/work_F/cwi-
code/omuse_env/share/amuse/lib/forsockets worker_code.f90 interface.o /home/fjansson/work_F/cwi-code/omuse/src/omuse/c
ommunity/dales/dales-build/libdales.a -L/home/fjansson/work_F/cwi-code/omuse_env/share/amuse/lib/forsockets -lforsocke
ts -lnetcdf  -lm -ldl -lnetcdff  -L/home/fjansson/work_F/cwi-code/omuse_env/share/amuse/lib/stopcond -lstopcondmpi -o 
dales_worker
worker_code.f90:139:0:

  139 | FUNCTION internal__accept_on_port(port_identifier, comm_identifier)
      | 
Error: Function result ‘internal__accept_on_port’ at (1) has no IMPLICIT type
worker_code.f90:2516:26:

 2516 |         integers_out(1) = internal__accept_on_port( &
      |                          1
Error: Cannot convert UNKNOWN to INTEGER(4) at (1)
worker_code.f90:836:27:

  783 |         call MPI_BCast(header_in, HEADER_SIZE , MPI_INTEGER, 0, MPI_COMM_WORLD, ioerror)
      |                       2    
......
  836 |             call MPI_BCast(longs_in, header_in(HEADER_LONG_COUNT), MPI_INTEGER8, 0, MPI_COMM_WORLD, ioError);
      |                           1
Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(8)/INTEGER(4)).
worker_code.f90:843:25:
```

I met MPI problem also in DALES itself, where it could be solved by adding
"use mpi" in the code or compiling with -fallow-argument-mismatch.

**Comment by fjansson** [comment]
I got it to work by adding `-std=legacy` to the gfortran compilation options for both DALES itself and for the worker_code. 

Additionally there is a typo worke_rcode.f90, in the return value declaration of the function 'internal__accept_on_port':
`INTEGER :: internal__accept_on_portq`
This is the cause of the first error message.

**Comment by ipelupessy** [comment]
ok, thanks! I think for the framework code we should fix the type errors.  These will also crop up for a lot of lecagy fortran code (currently these are warnings), so there the -std=legacy option is probably a good idea...

**Comment by ipelupessy** [comment]
I cannot find the typo though...

**Comment by fjansson** [comment]
I think I saw the error on two different machines, so I concluded I didn't add it myself. 
What I think is wrong is the 'q' at the end of the declaration of the return type: `INTEGER :: internal__accept_on_portq`. I see this line in worker_code.f90, but I don't know where it's generated.

**Comment by ipelupessy** [comment]
yea, i looked at the generating code (a teplate in rfi/tools), nor could i find it anywhere in a AMUSE install, so it is a bit mistifying...

**Comment by ipelupessy** [comment]
also: do you know which mpi was used? I cannot reproduce argument mismatch warnings with gfortran 9.3.1 -Wargument-mismatch option, I think it should reproduce these (unless the MPI  definitions are different...)

**Comment by fjansson** [comment]
I have openmpi 4.0.3. I think the error comes from calling an mpi function twice with different argument types, *without* a use mpi statement. From the error messages I guess the compiler infers the argument types from the first call and then thinks the second is wrong.

I can't find that typo either, maybe it was my mistake after all.

**Comment by ipelupessy** [comment]
can you check the mpif.h ? that should be included in the fortran worker_code.f90 included

**Comment by fjansson** [comment]
I have mpif.h, and it is included in worker_code. However, I see here https://www.mpi-forum.org/docs/mpi-3.1/mpi31-report/node411.htm that `use mpi` is preferred over including mpif.h, one of the reasons being argument checking. Also, I couldn't find any mention of MPI_BCast in mpif.h or the files it includes.

**Comment by fjansson** [comment]
And here https://github.com/cmbant/forutils/commit/6502f284422d82e4e657e66abbda2b91e6ef6224 I see someone else meeting the same problem with MPI_Bcast, mpif.h and gfortran 10. They solve it by adding an explicit interface rather than use mpi, (because of some .mod version problem?)

**Comment by ipelupessy** [comment]
I have created PR #649 - for me it resolves build issues..can you maybe check and test?
(though my guess there will bemaybe  some older gfortran/ MPI combinations that won't like this...)

**Comment by fjansson** [comment]
Your PR branch works for me with gfortran 10, I just tested with a pip install of amuse pointing to your branch, and then building DALES. I still had to patch DALES itself (-std=legacy in DALES own cmake file), but that's a separate problem.

DALES has had `use mpi` for a long time, and no include statements, and this has not given problems with older compilers or other MPI as far as I know.

**Comment by ipelupessy** [comment]
ok, I think if you use MPI and its not working now, you have a mismatch between the code and the function prototypes (e.g. there was 1 in the AMUSE interface code where a boolean mpi argument was given as an integer in the code, you probably want to fix that eventually).

**Comment by ipelupessy** [comment]
fixed by #649
